[Formatter] Filtering sub-selects causes syntax error

```sql
WITH contributors AS (SELECT 'Dan' AS name, 3 AS tenure),
toks AS (
    SELECT name FROM (
        SELECT * FROM contributors
        WHERE name = 'Dan'
    ) sub
    WHERE tenure = 3
)
SELECT * FROM toks;
```

raises

```
failed to query ...: near "WHERE": syntax error
--- FAIL: TestQuery (0.02s)
    --- FAIL: TestQuery/syntax_error_test (0.01s)
```


Adjusting the query to not use a filter in the sub-select parses and runs successfully i.e.

```sql
WITH contributors AS (SELECT 'Dan' AS name, 3 AS tenure),
toks AS (
    SELECT name FROM (
        SELECT * FROM contributors
    ) sub
    WHERE tenure = 3
)
SELECT * FROM toks;
```

--- a/internal/formatter.go
+++ b/internal/formatter.go
@@ -688,9 +688,9 @@ func (n *ColumnHolderNode) FormatSQL(ctx context.Context) (string, error) {
 }
 
 var tokensAfterFromClause = [...]string{"WHERE", "GROUP BY", "HAVING", "QUALIFY", "WINDOW", "ORDER BY", "COLLATE"}
+var removeExpressions := regexp.MustCompile(`\(.+?\)`)
 
 func (n *FilterScanNode) FormatSQL(ctx context.Context) (string, error) {
-	removeExpressions := regexp.MustCompile(`\(.+?\)`)
 	if n.node == nil {
 		return "", nil
 	}

--- a/internal/formatter.go
+++ b/internal/formatter.go
@@ -3,6 +3,7 @@ package internal
 import (
 	"context"
 	"fmt"
+	"regexp"
 	"strings"
 
 	"github.com/goccy/go-json"
@@ -686,7 +687,10 @@ func (n *ColumnHolderNode) FormatSQL(ctx context.Context) (string, error) {
 	return "", nil
 }
 
+var tokensAfterFromClause = [...]string{"WHERE", "GROUP BY", "HAVING", "QUALIFY", "WINDOW", "ORDER BY", "COLLATE"}
+
 func (n *FilterScanNode) FormatSQL(ctx context.Context) (string, error) {
+	removeExpressions := regexp.MustCompile(`\(.+?\)`)
 	if n.node == nil {
 		return "", nil
 	}
@@ -704,8 +708,17 @@ func (n *FilterScanNode) FormatSQL(ctx context.Context) (string, error) {
 			return fmt.Sprintf("%s HAVING %s", input, filter), nil
 		}
 	}
-	if strings.Contains(input, "WHERE") && input[len(input)-1] != ')' {
-		// expected to qualify clause
+	currentQuery := string(removeExpressions.ReplaceAllString(input, ""))
+
+	// Qualify the statement if the input is not wrapped in parens
+	queryWrappedInParens := currentQuery == ""
+	containsTokens := false
+	// or the input contains a token that would result in a syntax error
+	for _, token := range tokensAfterFromClause {
+		containsTokens = containsTokens || strings.Contains(currentQuery, token)
+	}
+
+	if !queryWrappedInParens && containsTokens {
 		return fmt.Sprintf("( %s ) WHERE %s", input, filter), nil
 	}
 	return fmt.Sprintf("%s WHERE %s", input, filter), nil

--- a/internal/formatter.go
+++ b/internal/formatter.go
@@ -713,7 +713,7 @@ func (n *FilterScanNode) FormatSQL(ctx context.Context) (string, error) {
 	// Qualify the statement if the input is not wrapped in parens
 	queryWrappedInParens := currentQuery == ""
 	containsTokens := false
-	// or the input contains a token that would result in a syntax error
+	// and the input contains a token that would result in a syntax error
 	for _, token := range tokensAfterFromClause {
 		containsTokens = containsTokens || strings.Contains(currentQuery, token)
 	}

--- a/internal/formatter.go
+++ b/internal/formatter.go
@@ -688,7 +688,7 @@ func (n *ColumnHolderNode) FormatSQL(ctx context.Context) (string, error) {
 }
 
 var tokensAfterFromClause = [...]string{"WHERE", "GROUP BY", "HAVING", "QUALIFY", "WINDOW", "ORDER BY", "COLLATE"}
-var removeExpressions := regexp.MustCompile(`\(.+?\)`)
+var removeExpressions = regexp.MustCompile(`\(.+?\)`)
 
 func (n *FilterScanNode) FormatSQL(ctx context.Context) (string, error) {
 	if n.node == nil {

--- a/query_test.go
+++ b/query_test.go
@@ -2781,6 +2781,35 @@ FROM Produce WHERE Produce.category = 'vegetable' QUALIFY rank <= 3`,
 				{"cabbage", int64(3)},
 			},
 		},
+		// Regression test goccy/go-zetasqlite#150
+		{
+			name: "qualify group",
+			query: `
+				WITH produce AS (
+					SELECT 'kale' AS item, 23 AS purchases
+				)
+				SELECT item, sum(purchases)
+				FROM produce
+				GROUP BY item
+				QUALIFY ROW_NUMBER() OVER (PARTITION BY item ORDER BY item) = 1
+			`,
+			expectedRows: [][]interface{}{{"kale", int64(23)}},
+		},
+		// Regression test goccy/go-zetasqlite#147
+		{
+			name: "subselect qualifier",
+			query: `
+				WITH produce AS (SELECT 'banana' AS item, 3 AS purchases),
+				toks AS (
+					SELECT item FROM (
+						SELECT * FROM produce
+						WHERE item = 'banana'
+					) sub
+					WHERE purchases = 3
+				)
+				SELECT * FROM toks;`,
+			expectedRows: [][]interface{}{{"banana"}},
+		},
 		{
 			name: "qualify direct",
 			query: `

